README.md
<h3 align="center">shinydisconnect</h3>
<h4 align="center">
  🔌 Show a nice message when a Shiny app disconnects or errors
  <br><br>
  <a href="https://daattali.com/shiny/shinydisconnect-demo/">Demo</a>
  &middot;
  by <a href="https://deanattali.com">Dean Attali</a>
</h4>

<p align="center">
  <a href="https://github.com/daattali/shinydisconnect/actions">
    <img src="https://github.com/daattali/shinydisconnect/workflows/R-CMD-check/badge.svg" alt="R build status" />
  </a> 
  <a href="https://cran.r-project.org/package=shinydisconnect">
    <img src="https://www.r-pkg.org/badges/version/shinydisconnect" alt="CRAN version" />
  </a>
</p>

---

A shiny app can disconnect for a variety of reasons: an unrecoverable error occurred in the app, the server went down, the user lost internet connection, or any other reason that might cause the shiny app to lose connection to its server.

{shinydisconnect} allows you to add a nice message to the user when the app disconnects.  The message works both locally (running Shiny apps within RStudio) and on Shiny servers (such as shinyapps.io, RStudio Connect, Shiny Server Open Source, Shiny Server Pro). See the [demo Shiny app](https://daattali.com/shiny/shinydisconnect-demo/) online for examples.

**Need Shiny help? [I'm available for consulting](https://attalitech.com/).**<br/>
**If you find {shinydisconnect} useful, please consider [supporting my work](https://github.com/sponsors/daattali)! ❤**

<p align="center">
  <a style="display: inline-block;" href="https://github.com/sponsors/daattali">
    <img height="35" src="https://i.imgur.com/034B8vq.png" />
  </a>
  <a style="display: inline-block;" href="https://paypal.me/daattali">
    <img height="35" src="https://camo.githubusercontent.com/0e9e5cac101f7093336b4589c380ab5dcfdcbab0/68747470733a2f2f63646e2e6a7364656c6976722e6e65742f67682f74776f6c66736f6e2f70617970616c2d6769746875622d627574746f6e40312e302e302f646973742f627574746f6e2e737667" />
  </a>
</p>

> This package is part of a larger ecosystem of packages with a shared vision: solving common Shiny issues and improving Shiny apps with minimal effort, minimal code changes, and clear documentation. Other packages for your Shiny apps:

| Package | Description | Demo |
|---|---|---|
| [shinyjs](https://deanattali.com/shinyjs/) | 💡 Easily improve the user experience of your Shiny apps in seconds | [🔗](https://deanattali.com/shinyjs/overview#demo) |
| [shinyalert](https://github.com/daattali/shinyalert/) | 🗯️ Easily create pretty popup messages (modals) in Shiny | [🔗](https://daattali.com/shiny/shinyalert-demo/) |
| [shinyscreenshot](https://github.com/daattali/shinyscreenshot/) | 📷 Capture screenshots of entire pages or parts of pages in Shiny apps | [🔗](https://daattali.com/shiny/shinyscreenshot-demo/) |
| [timevis](https://github.com/daattali/timevis/) | 📅 Create interactive timeline visualizations in R | [🔗](https://daattali.com/shiny/timevis-demo/) |
| [shinycssloaders](https://github.com/daattali/shinycssloaders/) | ⌛ Add loading animations to a Shiny output while it's recalculating | [🔗](https://daattali.com/shiny/shinycssloaders-demo/) |
| [colourpicker](https://github.com/daattali/colourpicker/) | 🎨 A colour picker tool for Shiny and for selecting colours in plots | [🔗](https://daattali.com/shiny/colourInput/) |
| [shinybrowser](https://github.com/daattali/shinybrowser/) | 🌐 Find out information about a user's web browser in Shiny apps | [🔗](https://daattali.com/shiny/shinybrowser-demo/) |
| [shinytip](https://github.com/daattali/shinytip/) | 💬 Simple flexible tooltips for Shiny apps | WIP |
| [shinymixpanel](https://github.com/daattali/shinymixpanel/) | 🔍 Track user interactions with Mixpanel in Shiny apps or R scripts | WIP |
| [shinyforms](https://github.com/daattali/shinyforms/) | 📝 Easily create questionnaire-type forms with Shiny | WIP |


# Table of contents

- [The problem](#problem)
- [Examples](#examples)
- [Installation](#install)
- [How to use](#usage)
- [Sponsors 🏆](#sponsors)

<h2 id="problem">The problem</h2>

Have you ever noticed how an error in your Shiny app looks very different when it happens locally (in RStudio on your laptop) compared to when it happens in production (in shinyapps.io or Shiny Server or Connect)? Locally, when a Shiny app breaks, you just get a grey screen. But when a deployed app breaks, you also get a little strip that says "Disconnected from server. Reload."

![shiny default message](inst/img/shinydisconnect-default-message.png)

You don't have any control over that message's text or position, and you don't have a way to get that message to appear both locally and in deployed apps.

Well, at least you didn't until now. {shinydisconnect} solves exactly these two issues, by allowing you to show a customized (and pretty!) message when a Shiny app disconnects or errors, regardless of where the app is running.

<h2 id="examples">Examples</h2>

For interactive examples and to see all the features, [check out the demo app](https://daattali.com/shiny/shinydisconnect-demo/).

**Example 1: basic usage ([code](inst/examples/basic/app.R))**

![basic screenshot](inst/img/basic.PNG)

**Example 2: using parameters ([code](inst/examples/advanced/app.R))**

![advanced screenshot](inst/img/advanced.PNG)

**Example 3: full-width and vertically centered ([code](inst/examples/special/app.R))**

You can also use `disconnectMessage2()` to get a similar message box to this one.

![special screenshot](inst/img/special.png)

**Example 4: hide the app**

This can be achieved by setting the opacity to 1 (fully opaque).

![special2 screenshot](inst/img/special2.PNG)


<h2 id="install">Installation</h2>

**For most users:** To install the stable CRAN version:

```r
install.packages("shinydisconnect")
```

**For advanced users:** To install the latest development version from GitHub:

```r
install.packages("remotes")
remotes::install_github("daattali/shinydisconnect")
```

<h2 id="usage">How to use</h2>

The package only has two functions: `disconnectMessage()` and `disconnectMessage2()`.

Call `disconnectMessage()` anywhere in a Shiny app's UI to add a nice message when a shiny app disconnects. `disconnectMessage()` has many parameters to modify the text, position, and colours of the disconnect message.

Note that it's not possible to distinguish between errors and timeouts - they will both show the same message.

Without using this package, a shiny app that disconnects will either just show a greyed out screen if running locally (with no message), or will show a small message in the bottom-left corner that you cannot modify when running in a server.

Basic usage:

```
ui <- fluidPage(
  disconnectMessage(),
  actionButton("disconnect", "Disconnect the app")
)
server <- function(input, output, session) {
  observeEvent(input$disconnect, {
    session$close()
  })
}
shinyApp(ui, server)
```

You can use `disconnectMessage2()` as a pre-configured version of `disconnectMessage()` with a specific set of parameters.

<h2 id="sponsors">

Sponsors 🏆

</h2>

> There are no sponsors yet

[Become the first sponsor for
{shinydisconnect}\!](https://github.com/sponsors/daattali/sponsorships?tier_id=39856)


NEWS.md
# shinydisconnect 0.1.1 (2023-08-19)

Add correct package file documentation (see r-lib/roxygen2#1491)

# shinydisconnect 0.1.0 (2020-07-14)

Initial CRAN release


R/disconnectMessage.R
#' Show a nice message when a shiny app disconnects or errors
#'
#' A shiny app can disconnect for a variety of reasons: an unrecoverable error occurred in
#' the app, the server went down, the user lost internet connection, or any other reason
#' that might cause the shiny app to lose connection to its server.\cr\cr
#' Call `disonnectMessage()`
#' anywhere in a Shiny app's UI to add a nice message when this happens. Works
#' locally (running Shiny apps within RStudio) and on Shiny servers (such as shinyapps.io,
#' RStudio Connect, Shiny Server Open Source, Shiny Server Pro).\cr\cr
#' See the [demo Shiny app](https://daattali.com/shiny/shinydisconnect-demo/) online for examples.\cr\cr
#' Note that it's not possible to distinguish between errors and timeouts - they will both
#' show the same message.
#'
#' You can also use [`disconnectMessage2()`] to use a pre-set combination of parameters
#' that produces a large centered message.
#'
#' @param text The text to show in the message.
#' @param refresh The text to show in a link that allows the user to refresh the page.
#' Use `refresh = ""` if you don't want to show a refresh link.
#' @param width The width of the message box. Must be either an integer, or the string
#' `"full"` to make the message take up the entire page width.
#' @param top The distance from the message to the top of the page. Must be either
#' an integer, or the string `"center"` to make the box vertically centered.
#' @param size The font size of the text. (integer).
#' @param background The background colour of the message box.
#' @param colour The colour of the text of the message box.
#' @param overlayColour The colour of the overlay to draw on the page behind the message box.
#' An overlay is used to "grey out" the application and draw attention to the message. Use
#' `overlayOpacity = 0` to disable the overlay.
#' @param overlayOpacity The opacity of the overlay, from 0 (fully transparent/not visible) to 1
#' (fully opaque). Use `overlayOpacity = 0` to disable the overlay.
#' @param refreshColour The colour of the refresh text link
#' @param css Any additional CSS rules to apply to the message box. For example,
#' `css = "padding: 0 !important; border: 3px solid red;"` will remove padding and add a border.
#' Note that you may need to use the `!important` rule to override default styles.
#' @examples
#' if (interactive()) {
#'   library(shiny)
#'   shinyApp(
#'     ui = fluidPage(
#'       disconnectMessage(),
#'       actionButton("disconnect", "Disconnect the app")
#'     ),
#'     server = function(input, output, session) {
#'       observeEvent(input$disconnect, {
#'         session$close()
#'       })
#'     }
#'   )
#' }
#' @export
disconnectMessage <- function(
  text = "An error occurred. Please refresh the page and try again.",
  refresh = "Refresh",
  width = 450,
  top = 50,
  size = 22,
  background = "white",
  colour = "#444444",
  overlayColour = "black",
  overlayOpacity = 0.6,
  refreshColour = "#337ab7",
  css = ""
) {

  checkmate::assert_string(text, min.chars = 1)
  checkmate::assert_string(refresh)
  checkmate::assert_numeric(size, lower = 0)
  checkmate::assert_string(background)
  checkmate::assert_string(colour)
  checkmate::assert_string(overlayColour)
  checkmate::assert_number(overlayOpacity, lower = 0, upper = 1)
  checkmate::assert_string(refreshColour)
  checkmate::assert_string(css)

  if (width == "full") {
    width <- "100%"
  } else if (is.numeric(width) && width >= 0) {
    width <- paste0(width, "px")
  } else {
    stop("disconnectMessage: 'width' must be either an integer, or the string \"full\".", call. = FALSE)
  }

  if (top == "center") {
    top <- "50%"
    ytransform <- "-50%"
  } else if (is.numeric(top) && top >= 0) {
    top <- paste0(top, "px")
    ytransform <- "0"
  } else {
    stop("disconnectMessage: 'top' must be either an integer, or the string \"center\".", call. = FALSE)
  }

  htmltools::tagList(
    getLocalTags(),
    htmltools::tags$head(
      htmltools::tags$style(
        glue::glue(
          .open = "{{", .close = "}}",

          "#shiny-disconnected-overlay { display: none !important; }",

          "#ss-overlay {
             background-color: {{overlayColour}} !important;
             opacity: {{overlayOpacity}} !important;
             position: fixed !important;
             top: 0 !important;
             left: 0 !important;
             bottom: 0 !important;
             right: 0 !important;
             z-index: 99998 !important;
             overflow: hidden !important;
             cursor: not-allowed !important;
          }",

          "#ss-connect-dialog {
             background: {{background}} !important;
             color: {{colour}} !important;
             width: {{width}} !important;
             transform: translateX(-50%) translateY({{ytransform}}) !important;
             font-size: {{size}}px !important;
             top: {{top}} !important;
             position: fixed !important;
             bottom: auto !important;
             left: 50% !important;
             padding: 0.8em 1.5em !important;
             text-align: center !important;
             height: auto !important;
             opacity: 1 !important;
             z-index: 99999 !important;
             border-radius: 3px !important;
             box-shadow: rgba(0, 0, 0, 0.3) 3px 3px 10px !important;
          }",

          "#ss-connect-dialog::before { content: '{{text}}' }",

          "#ss-connect-dialog label { display: none !important; }",

          "#ss-connect-dialog a {
             display: {{ if (refresh == '') 'none' else 'block' }} !important;
             color: {{refreshColour}} !important;
             font-size: 0 !important;
             margin-top: {{size}}px !important;
             font-weight: normal !important;
          }",

          "#ss-connect-dialog a::before {
            content: '{{refresh}}';
            font-size: {{size}}px;
          }",

          "#ss-connect-dialog { {{ htmltools::HTML(css) }} }"
        )
      )
    )
  )
}

#' Show a nice message when a shiny app disconnects or errors
#'
#' This function is a version of [`disconnectMessage()`] with a pre-set combination
#' of parameters that results in a large centered message.
#' @export
disconnectMessage2 <- function() {
  disconnectMessage(
    text = "Your session has timed out.",
    refresh = "",
    size = 70,
    colour = "white",
    background = "rgba(64, 64, 64, 0.9)",
    width = "full",
    top = "center",
    overlayColour = "#999",
    overlayOpacity = 0.7,
    css = "padding: 15px !important; box-shadow: none !important;"
  )
}


inst/examples/demo/server.R
function(input, output, session) {
  code <- reactive({
    code <- paste0(
      'disconnectMessage(\n',
      '  text = "', input$text, '",\n',
      '  refresh = "', if (input$show_refresh) input$refresh else "", '",\n',
      '  background = "', input$background, '",\n',
      '  colour = "', input$colour, '",\n'
    )

    if (input$show_refresh) {
      code <- paste0(
        code,
        '  refreshColour = "', input$refreshColour, '",\n'
      )
    }

    if (input$show_overlay) {
      code <- paste0(
        code,
        '  overlayColour = "', input$overlayColour, '",\n',
        '  overlayOpacity = ', input$overlayOpacity, ',\n'
      )
    } else {
      code <- paste0(
        code,
        '  overlayOpacity = 0,\n'
      )
    }

    if (input$full_width) {
      width <- "\"full\""
    } else {
      width <- input$width
    }

    if (input$center) {
      top <- "\"center\""
    } else {
      top <- input$top
    }

    code <- paste0(
      code,
      '  width = ', width, ',\n',
      '  top = ', top, ',\n',
      '  size = ', input$size, ',\n',
      '  css = "', input$css, '"\n',
      ')'
    )

    code
  })

  output$code <- renderText({
    paste0(
      "library(shiny)\nlibrary(shinydisconnect)\n\n",
      "ui <- fluidPage(\n  ",
      gsub("\n", "\n  ", code()),
      ",\n  actionButton('disconnect', 'Disconnect the app')\n)\n\n",
      "server <- function(input, output, session) {\n",
      "  observeEvent(input$disconnect, {\n    session$close()\n  })",
      "\n}\n\nshinyApp(ui, server)"
    )
  })

  observeEvent(input$show, {
    tag <- eval(parse(text = paste0("shinydisconnect::", code())))
    insertUI("body", "beforeEnd", tag, immediate = TRUE)
    session$close()
  })

  observeEvent(input$set1, {
    updateTextInput(session, "text", value = "An error occurred. Please refresh the page and try again.")
    updateCheckboxInput(session, "show_refresh", value = TRUE)
    updateTextInput(session, "refresh", value = "Refresh")
    colourpicker::updateColourInput(session, "background", value = "white")
    colourpicker::updateColourInput(session, "colour", value = "#444444")
    colourpicker::updateColourInput(session, "refreshColour", value = "#337ab7")
    updateSliderInput(session, "size", value = 22)
    updateCheckboxInput(session, "show_overlay", value = TRUE)
    colourpicker::updateColourInput(session, "overlayColour", value = "black")
    updateSliderInput(session, "overlayOpacity", value = 0.6)
    updateCheckboxInput(session, "full_width", value = FALSE)
    updateNumericInput(session, "width", value = 450)
    updateCheckboxInput(session, "center", value = FALSE)
    updateNumericInput(session, "top", value = 50)
    updateTextAreaInput(session, "css", value = "")
  })

  observeEvent(input$set2, {
    updateTextInput(session, "text", value = "Your session has timed out.")
    updateCheckboxInput(session, "show_refresh", value = FALSE)
    colourpicker::updateColourInput(session, "background", value = "rgba(64, 64, 64, 0.9)")
    colourpicker::updateColourInput(session, "colour", value = "white")
    colourpicker::updateColourInput(session, "refreshColour", value = "#337ab7")
    updateSliderInput(session, "size", value = 70)
    updateCheckboxInput(session, "show_overlay", value = TRUE)
    colourpicker::updateColourInput(session, "overlayColour", value = "#999")
    updateSliderInput(session, "overlayOpacity", value = 0.7)
    updateCheckboxInput(session, "full_width", value = TRUE)
    updateCheckboxInput(session, "center", value = TRUE)
    updateTextAreaInput(session, "css", value = "padding: 15px !important; box-shadow: none !important;")
  })

  observeEvent(input$set3, {
    updateTextInput(session, "text", value = "Something went wrong! Try refreshing the page.")
    updateCheckboxInput(session, "show_refresh", value = TRUE)
    colourpicker::updateColourInput(session, "background", value = "#000")
    colourpicker::updateColourInput(session, "colour", value = "#fff")
    colourpicker::updateColourInput(session, "refreshColour", value = "#337ab7")
    updateSliderInput(session, "size", value = 24)
    updateCheckboxInput(session, "show_overlay", value = TRUE)
    colourpicker::updateColourInput(session, "overlayColour", value = "#000")
    updateSliderInput(session, "overlayOpacity", value = 1)
    updateCheckboxInput(session, "full_width", value = FALSE)
    updateNumericInput(session, "width", value = 450)
    updateCheckboxInput(session, "center", value = TRUE)
  })
}


inst/examples/demo/ui.R
library(shiny)

share <- list(
  title = "{shinydisconnect} package",
  url = "https://daattali.com/shiny/shinydisconnect-demo/",
  image = "https://daattali.com/shiny/img/shinydisconnect.png",
  description = "Show a nice message when a shiny app disconnects or errors",
  twitter_user = "daattali"
)

fluidPage(
  title = paste0(share$title, " ", as.character(packageVersion("shinydisconnect"))),
  tags$head(
    includeCSS(file.path('www', 'style.css')),
    # Favicon
    tags$link(rel = "shortcut icon", type="image/x-icon", href="https://daattali.com/shiny/img/favicon.ico"),
    # Facebook OpenGraph tags
    tags$meta(property = "og:title", content = share$title),
    tags$meta(property = "og:type", content = "website"),
    tags$meta(property = "og:url", content = share$url),
    tags$meta(property = "og:image", content = share$image),
    tags$meta(property = "og:description", content = share$description),

    # Twitter summary cards
    tags$meta(name = "twitter:card", content = "summary"),
    tags$meta(name = "twitter:site", content = paste0("@", share$twitter_user)),
    tags$meta(name = "twitter:creator", content = paste0("@", share$twitter_user)),
    tags$meta(name = "twitter:title", content = share$title),
    tags$meta(name = "twitter:description", content = share$description),
    tags$meta(name = "twitter:image", content = share$image)
  ),
  tags$a(
    href="https://github.com/daattali/shinydisconnect",
    tags$img(style="position: absolute; top: 0; right: 0; border: 0;",
             src="github-green-right.png",
             alt="Fork me on GitHub")
  ),

  shinydisconnect::disconnectMessage2(),

  div(
    id = "header",
    div(id = "pagetitle", share$title),
    div(id = "subtitle", share$description),
    div(id = "subsubtitle",
        "Created by",
        tags$a(href = "https://deanattali.com/", "Dean Attali"),
        HTML("&bull;"),
        "Available",
        tags$a(href = "https://github.com/daattali/shinydisconnect", "on GitHub"),
        HTML("&bull;"),
        tags$a(href = "https://daattali.com/shiny/", "More apps"), "by Dean"
    )
  ),

  div(
    id = "main-row",
    fluidRow(column(
      width = 6, offset = 3,
      div(
        actionButton("show", "Simulate a disconnect", icon("play"), class = "btn-success"),
        br(), br()
      )
    )),
    fluidRow(column(
      width = 6, offset = 3,
      div(
        actionButton("set1", "Use configuration #1", class = "btn-primary"),
        actionButton("set2", "Use configuration #2", class = "btn-primary"),
        actionButton("set3", "Use configuration #3", class = "btn-primary")
      )
    ))
  ),

  fluidRow(
    column(
      3,
      h3("Text"),
      textInput("text", "Message", "An error occurred. Please refresh the page and try again."),
      checkboxInput("show_refresh", "Show a refresh button?", TRUE),
      conditionalPanel(
        "input.show_refresh",
        textInput("refresh", "Refresh text", "Refresh")
      ),
      sliderInput("size", "Font size", min = 10, max = 70, value = 22)
    ),
    column(
      3,
      h3("Style"),
      colourpicker::colourInput("background", "Message background", "white", showColour = "background", allowTransparent = TRUE),
      colourpicker::colourInput("colour", "Message text colour", "#444444", showColour = "background"),
      conditionalPanel(
        "input.show_refresh",
        colourpicker::colourInput("refreshColour", "Refresh text colour", "#337ab7", showColour = "background")
      ),
      textAreaInput("css", "Custom CSS (advanced)", "")
    ),
    column(
      3,
      h3("Overlay"),
      checkboxInput("show_overlay", "Add an overlay on the page?", TRUE),
      conditionalPanel(
        "input.show_overlay",
        colourpicker::colourInput("overlayColour", "Overlay colour", "black", showColour = "background"),
        sliderInput("overlayOpacity", "Overlay transparency", min = 0, max = 1, value = 0.6)
      )
    ),
    column(
      3,
      h3("Position"),
      checkboxInput("full_width", "Full width?", FALSE),
      conditionalPanel(
        "!input.full_width",
        numericInput("width", "Width", min = 0, value = 450)
      ),
      checkboxInput("center", "Vertically centered?", FALSE),
      conditionalPanel(
        "!input.center",
        numericInput("top", "Distance from top of page", min = 0, value = 50)
      )
    )
  ),

  fluidRow(
    column(
      12,
      h3("Generated code"),
      verbatimTextOutput("code")
    )
  )
)
